control/control.go

```diff
@@ -886,22 +921,79 @@ func (p *pluginControl) GetMetricVersions(ns core.Namespace) ([]core.CatalogedMe
 	return rmts, nil
 }
 
-func (p *pluginControl) MetricExists(mns core.Namespace, ver int) bool {
-	_, err := p.metricCatalog.Get(mns, ver)
+func (p *pluginControl) MetricsExists(mns core.Namespace, ver int) bool {
+	_, err := p.metricCatalog.GetMetric(mns, ver)
 	if err == nil {
 		return true
 	}
 	return false
 }
 
+// gatherMetricTypesGroupedByPlugin gathers matched metrics types for requested metrics and groups them by plugins to call
+func (p *pluginControl) gatherMetricTypesGroupedByPlugin(rmts []core.RequestedMetric, cdt *cdata.ConfigDataTree) (map[string]metricTypes, []error) {
+	pmts := make(map[string]metricTypes)
+	var errs []error
+
+	if cdt == nil {
+		// initialize config tree if nil is provided
+		cdt = cdata.NewTree()
+	}
+
+	for _, rmt := range rmts {
+		// for each requested metric get matched metrics types on appropriate version
+		// (if version < 1 choose the latest)
+		mts, err := p.metricCatalog.GetMatchedMetricTypes(rmt.Namespace(), rmt.Version())
+
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		for _, mt := range mts {
+			lp := mt.Plugin
+			if lp == nil {
+				errs = append(errs, errorMetricNotFound(mt.Namespace().Strings(), mt.Version()))
+				continue
+			}
+
+			// get config
+			config := cdt.Get(mt.Namespace().Strings())
+
+			if config == nil {
+				config = cdata.NewNode()
+			}
+			returnedmt := plugin.MetricType{
+				Namespace_:          mt.Namespace(),
+				LastAdvertisedTime_: mt.LastAdvertisedTime(),
+				Version_:            mt.Version(),
+				Tags_:               mt.Tags(),
+				Config_:             config,
+				Unit_:               mt.Unit(),
+			}
+
+			// group matched metrics by plugin
+			pluginKey := lp.Key()
+			pmt, _ := pmts[pluginKey]
+			pmt.plugin = lp
+			pmt.metricTypes = append(pmt.metricTypes, returnedmt)
+			pmts[pluginKey] = pmt
+
+		}
+
+	}
+
+	return pmts, errs
+}
+
 // CollectMetrics is a blocking call to collector plugins returning a collection
 // of metrics and errors.  If an error is encountered no metrics will be
 // returned.
-func (p *pluginControl) CollectMetrics(metricTypes []core.Metric, deadline time.Time, taskID string) (metrics []core.Metric, errs []error) {
+func (p *pluginControl) CollectMetrics(rmts []core.RequestedMetric, cdt *cdata.ConfigDataTree, deadline time.Time, taskID string) (metrics []core.Metric, errs []error) {
 
-	pluginToMetricMap, err := groupMetricTypesByPlugin(p.metricCatalog, metricTypes)
-	if err != nil {
-		errs = append(errs, err)
+	// gathers metrics to collect grouped by plugin
+	pluginToMetricMap, errs := p.gatherMetricTypesGroupedByPlugin(rmts, cdt)
+
+	if len(errs) > 0 {
```

Comment:
As you can see at https://github.com/intelsdi-x/snap/blob/master/control/control.go#L900 I kept in that form as it was before.